File: Geography/src/geography/agents/Shelter.java
package geography.agents;

import geography.routing.StreetNetwork;

/**
 * A clean-air (smoke-respite) shelter.
 *
 * Real September 2020 shelters are loaded from
 * {@code data/shelters/shelters_2020-09.csv} (provenance in
 * Geography/data/README.md and docs/science/DATA_SOURCES.md D1). Each shelter
 * is snapped to its nearest street-graph node at context build and carries the
 * Dijkstra shortest-path tree rooted there, so every resident can look up its
 * true network distance and reconstruct a walking path
 * ({@link geography.routing.StreetNetwork}).
 *
 * Capacity is enforced: a shelter admits residents up to {@link #capacity};
 * once full it refuses admission and the resident must re-route (V12,
 * docs/science/DESIGN_SPEC.md). A null capacity means "not capacity-limited"
 * (used for standby sites that are modelled as always-available if operating).
 */
public class Shelter {
    private final String id;
    private final String name;
    /** Nightly capacity; null = not capacity-limited. Provenance: see DATA_SOURCES D1. */
    private final Integer capacity;
    /** Whether this site is part of the active scenario (standby sites are off by default). */
    private final boolean operating;
    private final double lon;
    private final double lat;

    /** Tick this shelter opens / closes, from the real `opened`/`closed` dates in
     *  the shelter CSV (D1: OCC 2020-09-10, CJ 2020-09-11, both to 2020-09-19).
     *  Set by ContextCreator relative to the simulation start. Negative infinity
     *  / positive infinity when the opening-date gate is disabled, which
     *  reproduces the always-open behaviour of every run before this commit.
     *
     *  <p><b>Assumption (documented, A-02 mitigation).</b> The source gives a
     *  DATE, not an hour, so opening is taken as 00:00 local on that date — the
     *  earliest defensible reading. Real shelters opened later in the day, so
     *  this UNDERSTATES pre-opening outdoor exposure; the direction of the bias
     *  is known and stated rather than tuned away. */
    private double openTick = Double.NEGATIVE_INFINITY;
    private double closeTick = Double.POSITIVE_INFINITY;

    /** Street-graph node this shelter is snapped to (RLIS node id). */
    private long graphNodeId = -1;
    /** Dijkstra tree rooted at {@link #graphNodeId}; set once at context build. */
    private StreetNetwork.ShortestPathTree routeTree;

    /** Spaces held back for priority (mobility-limited) arrivals under the
     *  need-based-admission scenario (arm D). Set by ContextCreator from the
     *  {@code triageReserveFraction} parameter, clamped to [0, capacity].
     *
     *  <p><b>Zero by default, and zero is exactly the old behaviour.</b> With
     *  {@code reservedForPriority == 0} the priority-aware predicates below are
     *  arithmetically identical to the first-come-first-served ones they
     *  replaced, so arms A/B/C reproduce byte-identically. */
    private int reservedForPriority = 0;

    /** Number of residents currently admitted (V12 occupancy). */
    private int occupancy = 0;
    /** Peak occupancy over the run (shelter-level export metric). */
    private int peakOccupancy = 0;
    /** Count of admission refusals due to capacity (shelter-level export metric). */
    private int refusedCount = 0;

    public Shelter(String id, String name, Integer capacity, boolean operating, double lon, double lat) {
        this.id = id;
        this.name = name;
        this.capacity = capacity;
        this.operating = operating;
        this.lon = lon;
        this.lat = lat;
    }

    /**
     * Attempts to admit one resident. Returns true and increments occupancy if
     * there is room (or the shelter is not capacity-limited); otherwise records
     * a refusal and returns false.
     */
    public boolean admit() {
        return admit(false);
    }

    /**
     * Attempts to admit one resident under need-based admission (arm D).
     * A priority arrival may use the whole capacity; a non-priority arrival is
     * refused once occupancy reaches {@code capacity - reservedForPriority}.
     * With the reserve at 0 this is identical to {@link #admit()}.
     */
    public boolean admit(boolean isPriority) {
        if (!hasSpaceFor(isPriority)) {
            refusedCount++;
            return false;
        }
        occupancy++;
        if (occupancy > peakOccupancy) {
            peakOccupancy = occupancy;
        }
        return true;
    }

    /** True if this shelter can currently admit at least one more resident. */
    public boolean hasSpace() {
        return hasSpaceFor(false);
    }

    /** True if this shelter can currently admit one more resident of the given
     *  priority class. Non-priority residents see only the unreserved part of
     *  the capacity; priority residents see all of it. */
    public boolean hasSpaceFor(boolean isPriority) {
        if (capacity == null) {
            return true;
        }
        int usable = isPriority ? capacity : capacity - reservedForPriority;
        return occupancy < usable;
    }

    /** Sets the number of spaces held for priority arrivals, clamped into
     *  [0, capacity]. No-op for capacity-unlimited shelters. */
    public void setReservedForPriority(int reserved) {
        if (capacity == null) {
            this.reservedForPriority = 0;
            return;
        }
        this.reservedForPriority = Math.max(0, Math.min(capacity.intValue(), reserved));
    }

    public int getReservedForPriority() { return reservedForPriority; }

    /** True if this shelter is physically open at the given tick. Always true
     *  when the opening-date gate is disabled (openTick = -inf, closeTick = +inf). */
    public boolean isOpenAt(double tick) {
        return tick >= openTick && tick < closeTick;
    }

    /** True if this shelter can be selected and entered right now: operating in
     *  the scenario, open on the calendar, and not yet full. */
    public boolean isAvailableAt(double tick) {
        return isAvailableAt(tick, false);
    }

    /** As {@link #isAvailableAt(double)} but for a resident of the given
     *  priority class (arm D). Identical when the reserve is 0. */
    public boolean isAvailableAt(double tick, boolean isPriority) {
        return operating && isOpenAt(tick) && hasSpaceFor(isPriority);
    }

    public void setOpenWindowTicks(double openTick, double closeTick) {
        this.openTick = openTick;
        this.closeTick = closeTick;
    }

    public double getOpenTick() { return openTick; }
    public double getCloseTick() { return closeTick; }

    public String getId() { return id; }
    public String getName() { return name; }
    public Integer getCapacity() { return capacity; }
    public boolean isOperating() { return operating; }
    public double getLon() { return lon; }
    public double getLat() { return lat; }
    public int getOccupancy() { return occupancy; }
    public int getPeakOccupancy() { return peakOccupancy; }
    public int getRefusedCount() { return refusedCount; }

    public long getGraphNodeId() { return graphNodeId; }
    public void setGraphNodeId(long graphNodeId) { this.graphNodeId = graphNodeId; }
    public StreetNetwork.ShortestPathTree getRouteTree() { return routeTree; }
    public void setRouteTree(StreetNetwork.ShortestPathTree routeTree) { this.routeTree = routeTree; }

    @Override
    public String toString() { return id; }
}

File: websim/engine/src/shelters/shelter.ts
/**
 * `Shelter` — a clean-air (smoke-respite) site.
 *
 * Verbatim port of `geography/agents/Shelter.java`. Four semantics here are
 * load-bearing and each has burned someone before:
 *
 * 1. **`capacity === null` means UNLIMITED**, not zero. It comes from a blank
 *    `capacity` cell, and `hasSpaceFor` returns `true` unconditionally for such
 *    a site. A port that coerced blank → 0 would turn standby sites into closed
 *    ones and every occupancy number downstream would be wrong in the safe-looking
 *    direction.
 * 2. **`admit()` mutates on failure.** A refused admission increments
 *    `refusedCount`. It must therefore never be called speculatively — the
 *    caller asks `hasSpaceFor` when it wants a question answered and `admit`
 *    only when it means to walk through the door (PORT_MAP §1.5 step 13).
 * 3. **The triage reserve is a floor, not a round** (`ContextCreator.java:567`),
 *    and it is applied only when `capacity != null && fraction > 0`. At fraction
 *    0 the reserve is 0 and `hasSpaceFor` is arithmetically the expression it was
 *    before arm D existed — which is why arms A/B/C reproduce bit-identically.
 * 4. **`isOpenAt` is `tick >= openTick && tick < closeTick`** — closed-inclusive
 *    on the left, exclusive on the right, over `double` ticks that are ±Infinity
 *    when the opening-date gate is off.
 *
 * `petIntake` is a **tri-state**: `true` admit, `false` refuse, `null`
 * unrecorded. Only `null` defers to the run-wide `petPolicyDefault`. Collapsing
 * it to a boolean silently converts every unrecorded site into whichever default
 * the port's author preferred.
 */

import type { ShortestPathTree } from "../graph/dijkstra.js";

export class Shelter {
  readonly id: string;
  readonly name: string;
  /** Nightly capacity; `null` = not capacity-limited. */
  readonly capacity: number | null;
  /** Part of the active scenario (standby sites are off by default). */
  readonly operating: boolean;
  readonly lon: number;
  readonly lat: number;

  /** ±Infinity when `respectShelterOpeningDates` is 0 — i.e. always open. */
  private openTickValue = Number.NEGATIVE_INFINITY;
  private closeTickValue = Number.POSITIVE_INFINITY;

  /** Node **index** into the routing graph (not the sparse certified node id). */
  graphNode = -1;
  /** Certified node id of {@link graphNode}; what fixtures and outputs carry. */
  graphNodeId = -1;
  /** Geodesic displacement introduced by snapping, in metres (build-time). */
  snapGapM = Number.NaN;
  /** Dijkstra tree rooted at {@link graphNode}; recomputed after each closure wave. */
  routeTree: ShortestPathTree | null = null;

  private reserved = 0;
  private occupancyValue = 0;
  private peakOccupancyValue = 0;
  private refusedCountValue = 0;
  private policyRefusedCountValue = 0;

  /** Tri-state pet policy: `true` admit, `false` refuse, `null` unrecorded. */
  petIntake: boolean | null = null;
  adultsOnly = false;

  constructor(
    id: string,
    name: string,
    capacity: number | null,
    operating: boolean,
    lon: number,
    lat: number,
  ) {
    this.id = id;
    this.name = name;
    this.capacity = capacity;
    this.operating = operating;
    this.lon = lon;
    this.lat = lat;
  }

  /**
   * Attempts to admit one resident.
   *
   * **Side-effecting on failure**: a refusal increments `refusedCount`. See
   * class doc point 2.
   */
  admit(isPriority = false): boolean {
    if (!this.hasSpaceFor(isPriority)) {
      this.refusedCountValue++;
      return false;
    }
    this.occupancyValue++;
    if (this.occupancyValue > this.peakOccupancyValue) {
      this.peakOccupancyValue = this.occupancyValue;
    }
    return true;
  }

  /**
   * Can this site take one more resident of the given priority class?
   *
   * Non-priority arrivals see `capacity − reservedForPriority`; priority
   * arrivals (mobility-limited only — never age/asthma/COPD) see all of it.
   */
  hasSpaceFor(isPriority: boolean): boolean {
    if (this.capacity === null) {
      return true;
    }
    const usable = isPriority ? this.capacity : this.capacity - this.reserved;
    return this.occupancyValue < usable;
  }

  hasSpace(): boolean {
    return this.hasSpaceFor(false);
  }

  /** Clamped into `[0, capacity]`; a no-op for capacity-unlimited sites. */
  setReservedForPriority(reserved: number): void {
    if (this.capacity === null) {
      this.reserved = 0;
      return;
    }
    this.reserved = Math.max(0, Math.min(this.capacity, reserved));
  }

  get reservedForPriority(): number {
    return this.reserved;
  }

  setOpenWindowTicks(openTick: number, closeTick: number): void {
    this.openTickValue = openTick;
    this.closeTickValue = closeTick;
  }

  get openTick(): number {
    return this.openTickValue;
  }

  get closeTick(): number {
    return this.closeTickValue;
  }

  isOpenAt(tick: number): boolean {
    return tick >= this.openTickValue && tick < this.closeTickValue;
  }

  /** Operating in the scenario, open on the calendar, and not yet full. */
  isAvailableAt(tick: number, isPriority = false): boolean {
    return this.operating && this.isOpenAt(tick) && this.hasSpaceFor(isPriority);
  }

  /**
   * A policy refusal (pet / adults-only gate). Counted into `refusedCount` so
   * total refusals stay one number, and separately so the share is auditable.
   */
  recordPolicyRefusal(): void {
    this.refusedCountValue++;
    this.policyRefusedCountValue++;
  }

  get occupancy(): number {
    return this.occupancyValue;
  }

  get peakOccupancy(): number {
    return this.peakOccupancyValue;
  }

  get refusedCount(): number {
    return this.refusedCountValue;
  }

  get policyRefusedCount(): number {
    return this.policyRefusedCountValue;
  }

  toString(): string {
    return this.id;
  }
}
